### PR TITLE
Add monitor-dog

### PR DESCRIPTION
**Pull request by podviaznikov**
Add counter and timer for all ponos jobs. Hidden behind flag.

- increment for all started jobs
- increment for all jobs results: success, timeout error, fatal error, task error
- ponos.timer to track how long each task execution took

Tasks:

- [x] please merge https://github.com/Runnable/monitor-dog/pull/11 first
- [x] use merged version of monitor-dog
- [x] review
- [ ] demo

### CODE REVIEW

**Comment by podviaznikov** [comment]
@anandkumarpatel, @bkendall, @rsandor please review

File: lib/worker.js [hunk 1]
```diff
@@ -121,6 +141,10 @@ Worker.prototype.run = function () {
     queue: this.queue,
     job: this.job
   }
+  if (process.env.WORKER_MONITOR) {
```
Comment (anandkumarpatel) [inline lib/worker.js]: I don't think for us there is a case where we don't want to get this data. Might be better to check for DATADOG_HOST

Comment (anandkumarpatel) [inline lib/worker.js]: Well at that point you don't need to check that here monitordog already does that

File: lib/worker.js [hunk 2]
```diff
@@ -111,6 +112,25 @@ Worker.prototype._reportError = function (err) {
 }
 
 /**
+ * Helper function for creating monitor-dog events tags
+ * @private
+ * @param {Object} tags as Object {queue: 'docker.event.publish'}
```
Comment (bkendall) [inline lib/worker.js]: `@param` -> `@returns` ?

Comment (bkendall) [inline lib/worker.js]: can you expand this jsdoc to illustrate what it's doing? I _think_ it's returning the queue name as a value on an object, but the reverse in there makes me curious...

File: test/unit/worker.js [hunk 3]
```diff
@@ -150,12 +151,56 @@ describe('Worker', function () {
...
+      assert.isObject(tags)
+      assert.equal(Object.keys(tags).length, 4)
+      assert.deepEqual(tags, {
+        queue: queue,
+        token0: 'name',
+        token1: 'queue.name',
+        token2: 'some.queue.name'
+      })
+    })
+    it('should generate tags for old style queues', function () {
```
Comment (bkendall) [inline test/unit/worker.js]: newline before these two `it`s (see above block for example)

File: test/unit/worker.js [hunk 4]
```diff
@@ -193,6 +277,46 @@ describe('Worker', function () {
...
+              .then(function () {
+                // this is asserting taskHandler called once, but was twice
+                assert.ok(taskHandler.calledOnce, 'task was called once')
+                assert.ok(doneHandler.calledOnce, 'done was called once')
+                sinon.assert.callCount(monitor.increment, 5)
+                sinon.assert.calledWith(monitor.increment.firstCall, 'ponos', {
+                  token0: 'command',
+                  token1: 'something.command',
+                  token2: 'do.something.command',
+                  queue: 'do.something.command' })
```
Comment (bkendall) [inline test/unit/worker.js]: closing bracket and parenthesis go on a new line

File: lib/worker.js [hunk 5]
```diff
@@ -143,17 +167,32 @@ Worker.prototype.run = function () {
         merge(jobAndQueueData, { result: result }),
         'Task complete'
       )
+      if (process.env.WORKER_MONITOR) {
+        var tags = this._eventTags()
+        tags.result = 'success'
+        monitor.increment('ponos.finish', tags)
+      }
```
Comment (bkendall) [inline lib/worker.js]: can you pull these `if (process.env.WORKER_MONITOR)` blocks out into some helper? It's pretty much the same block repeated multiple times, so you should be able to pull it out. Then the code isn't littered with `if`s as well

Comment (bkendall) [inline lib/worker.js]: @anandkumarpatel and I just had a little discussion. What do you think about doing something like:

```js
if (process.env.WORKER_MONITOR_DISABLED) {
  return
}
// do monitoring
```

Then it allows us to enable it via `HOST` and `PORT` environment variables, but still allows us to disable it if necessary (UDPocolypse or something)

Comment (anandkumarpatel) [inline lib/worker.js]: +1

Comment (anandkumarpatel) [inline lib/worker.js]: I agree, make this negative check, that way we do not have to add extra var in all our projects configs.

Comment (podviaznikov) [inline lib/worker.js]: good idea. I like it

File: test/unit/worker.js (hunk 3 above)
Comment (podviaznikov) [inline test/unit/worker.js]: done

File: lib/worker.js (hunk 5 above)
Comment (podviaznikov) [inline lib/worker.js]: done

File: lib/worker.js (hunk 2 above)
Comment (podviaznikov) [inline lib/worker.js]: updated docs

File: lib/worker.js [hunk 6]
```diff
@@ -111,6 +112,51 @@ Worker.prototype._reportError = function (err) {
 }
 
 /**
+ * Helper function for creating monitor-dog events tags
+ * `queue` is the only mandatory tag.
+ * Few tags would be created depednding on the queue name
```
Comment (bkendall) [inline lib/worker.js]: `depednding` -> `depending`

File: test/unit/worker.js (hunk 3 above)
Comment (bkendall) [inline test/unit/worker.js]: Not quite. Still needs some checking in this file.

File: lib/worker.js [hunk 7]
```diff
@@ -111,6 +112,51 @@ Worker.prototype._reportError = function (err) {
...
+}
+
+/**
+ * Helper function calling `monitor.increment`.
+ * Monitor wouldn't be called if `process.env.WORKER_MONITOR_DISABLED` set.
+ * @param {string} eventName name to be reported into the datadog
+ * @param {object} [extraTags] extra tags to be send with the event
+ * @private
+ */
+Worker.prototype._inc = function (eventName, extraTags) {
```
Comment (bkendall) [inline lib/worker.js]: Can we rename this function? `_inc` is maddeningly unhelpful when I was reading it below. `_incMonitor`, maybe?

Comment (podviaznikov) [inline lib/worker.js]: good suggestion. Had trouble coming up with the good name in this case

File: lib/worker.js (hunk 6 above)
Comment (podviaznikov) [inline lib/worker.js]: good catch

**Comment by podviaznikov** [comment]
@bkendall one more review, please

File: lib/worker.js [hunk 8]
```diff
@@ -121,6 +167,8 @@ Worker.prototype.run = function () {
     queue: this.queue,
     job: this.job
   }
+  this._incMonitor('ponos')
+  var timer = monitor.timer('ponos.timer', !process.env.WORKER_MONITOR_DISABLED, this._eventTags())
```
Comment (bkendall) [inline lib/worker.js]: what is `!process.env.WORKER_MONITOR_DISABLED` here for?

Comment (podviaznikov) [inline lib/worker.js]: good questions @bkendall. So second parameter to `monitor.timer` is `start`. If it false timer will not be started.
`timer.stop` will also work correctly.
Using this flag I don't need to add additional `if` around `monitor.timer`.

File: test/unit/worker.js [hunk 9]
```diff
@@ -166,9 +267,61 @@ describe('Worker', function () {
...
+              assert.ok(doneHandler.calledOnce, 'done was called once')
+              sinon.assert.notCalled(monitor.increment)
+              sinon.assert.calledOnce(monitor.timer)
+              sinon.assert.calledWith(monitor.timer, 'ponos.timer',
+                !process.env.WORKER_MONITOR_DISABLED, {
+                  token0: 'command',
+                  token1: 'something.command',
+                  token2: 'do.something.command',
+                  queue: 'do.something.command'
+                })
```
Comment (bkendall) [inline test/unit/worker.js]: either get everything on one line, or put each argument on new lines.

i.e.:

```js
// bad:
assert(foo, bar
  baz, {
    value: 4
  })

// good (the last object started on the same line,
//   which means it closes on the same column as the assert):
assert(foo, bar baz, {
  value: 4
})

// good (new lines all around):
assert(
  foo,
  bar
  baz,
  {
    value: 4
  }
)
```

(sorry to be picky, but I really hate multiple lines with arbitrary numbers of arguments on each line)

File: test/unit/worker.js [hunk 10]
```diff
@@ -191,6 +344,29 @@ describe('Worker', function () {
               // this is asserting taskHandler called once, but was twice
               assert.ok(taskHandler.calledOnce, 'task was called once')
               assert.ok(doneHandler.calledOnce, 'done was called once')
+              sinon.assert.callCount(monitor.increment, 5)
+              sinon.assert.calledWith(monitor.increment.firstCall, 'ponos', {
+                token0: 'command',
+                token1: 'something.command',
+                token2: 'do.something.command',
+                queue: 'do.something.command'
+              })
```
Comment (bkendall) [inline test/unit/worker.js]: to the earlier point about things on lines: this sample assert block is fine - everything (including the start of the object) is on one line, so I know that that's all the arguments, and the last one's an object (which is closed in-line with the `sinon.assert`). Make scanning through code so much easier

File: test/unit/worker.js [hunk 11]
```diff
@@ -191,6 +344,29 @@ describe('Worker', function () {
...
+              sinon.assert.calledWith(monitor.increment.secondCall, 'ponos.finish', {
+                result: 'timeout-error',
+                token0: 'command',
+                token1: 'something.command',
+                token2: 'do.something.command',
+                queue: 'do.something.command'
+              })
+              sinon.assert.calledTwice(monitor.timer)
+              sinon.assert.calledWith(monitor.timer, 'ponos.timer',
+                !process.env.WORKER_MONITOR_DISABLED, {
```
Comment (bkendall) [inline test/unit/worker.js]: then this sneaky `process.env` shows up again. I keep thinking it's a value in some object that is defined, but it's an entirely new argument!

Comment (podviaznikov) [inline test/unit/worker.js]: I think I got it now

File: test/unit/worker.js (hunk 9 above)
Comment (podviaznikov) [inline test/unit/worker.js]: I think I've got it right this time

File: lib/worker.js (hunk 8 above)
Comment (bkendall) [inline lib/worker.js]: > Optional boolean startNow can be provided to start the timer at a later date using the .start method.

The way I'm reading it, if this is set to `true`, then the timer will never be started, but it is created. My worry here is that all the workers will create timers that are never started and they'll all be stored in memory, leading to a memory leak. I would be okay with this block being wrapped with an `if` statement, or wrapping it in a new function like you did with the increment. You'd end up with

```js
Worker prototype.createTimer () { return !process.env.WORKER_MONITOR_DISABLED ? monitor.timer() : null }
 
var timer = this.createTimer()
// ...
if (timer) { timer.stop() }
```

**Comment by bkendall** [comment]
one cosmetic thing and one leaky thing, and then I think you're good to go! :smile:

File: lib/worker.js (hunk 8 above)
Comment (podviaznikov) [inline lib/worker.js]: will do this. This suggestion is nice

**Comment by bkendall** [comment]
LGTM! I'm going to run a test or two locally, but looks very good!

**Pull request merged**